chainimport/iterators.go

```diff
@@ -0,0 +1,160 @@
+package chainimport
+
+import (
+	"io"
+	"iter"
+)
+
+// importSourceHeaderIterator provides efficient iteration over headers from
+// a source.
+type importSourceHeaderIterator struct {
+	// source is the HeaderImportSource from which headers are retrieved
+	// during iteration.
+	source HeaderImportSource
+
+	// currentIndex tracks the current position in the iteration sequence,
+	// representing the next header index to be processed.
+	currentIndex uint32
+
+	// endIndex specifies the final header index (inclusive) that will be
+	// processed by this iterator.
+	endIndex uint32
+
+	// batchSize determines how many headers are fetched in a single
+	// operation to optimize performance during iteration.
+	batchSize uint32
+}
+
+// Compile-time assertion to ensure importSourceHeaderIterator implements
+// headerIterator interface.
+var _ HeaderIterator = (*importSourceHeaderIterator)(nil)
+
+// Iterator returns a stateless iter.Seq2[header, error] for the specified
+// headers range. Each call creates a fresh iterator that independently
+// traverses the range.
+func (it *importSourceHeaderIterator) Iterator(startIdx,
+	endIdx uint32) iter.Seq2[Header, error] {
+
+	return func(yield func(Header, error) bool) {
+		for idx := startIdx; idx <= endIdx; idx++ {
+			header, err := it.source.GetHeader(idx)
+			if err != nil {
+				if !yield(nil, err) {
+					return
+				}
+				return
+			}
+
+			if header != nil {
+				if !yield(header, nil) {
+					return
+				}
+			}
+		}
+	}
+}
+
+// BatchIterator returns a stateless iterator that yields batches of headers for
+// the specified range, where each batch respects the configured batch size
+// limit. Each call creates a fresh iterator that independently traverses the
+// range.
+func (it *importSourceHeaderIterator) BatchIterator(startIdx,
+	endIdx uint32) iter.Seq2[[]Header, error] {
+
+	return func(yield func([]Header, error) bool) {
+		currentIdx := startIdx
+		for currentIdx <= endIdx {
+			batch, err := it.ReadBatch(currentIdx, endIdx)
+			if err != nil {
+				if !yield(nil, err) {
+					return
+				}
+				return
+			}
+
+			if len(batch) == 0 {
+				return
+			}
+
+			if !yield(batch, nil) {
+				return
+			}
+
+			currentIdx += uint32(len(batch))
+		}
+	}
+}
+
+// Next returns the next header in the current stateful iteration sequence. It
+// maintains internal position state. Returns nil, io.EOF when no more headers
+// are available.
+func (it *importSourceHeaderIterator) Next() (Header, error) {
```

Comment:
Not super blocking, but I think we can enable this to function as a stateful iterator by making another method name like `StateFullIter` or something. Then similar to the function above, but we keep `currentIndex` and `endIndex` as closure variables. Then someone can use it as normal, or with `iter.Pull`, etc.